### PR TITLE
SAMZA-2741: [Pipeline Drain] Fix processing of Drain messages for High-Level and Low-level API

### PR DESCRIPTION
**Pull request by ajothomas**
# Symptoms and Cause:
There were a few issues with the processing of drain messages that this PR attempts to fix. The following issues were encountered during the end-to-end testing with test pipelines.

1. It was observed that the `RunLoop` was running into an exception when a task was trying to consume from intermediate streams.

This was happening as we stop all consumers in `SystemConsumers` when we encounter drain. This also stops the consumers which ingest data from the intermediate streams. 

_After 1 was resolved, some additional issues were observed-_
2. The pipeline was getting stuck after partial consumption of data after the last shuffle stage. 

This was happening as we were removing the SSPs from the processing set of SSPs for a task for high-level API processing. This was leading to unprocessed drain messages and thereby, the pipeline would remain stuck.

3. The pipeline would get stuck if a drain message was encountered prior to the start of the container. 

This is happening as we set the `SystemConsumers` to drain mode before it is even started by `SamzaContainer`. As a result, no messages are read by the container.

# Changes:
1. Changed the `SystemConsumers` code to remove the line which stopped registered `SystemConsumer`s on drain. 

2. Restrict the logic to remove the SSP from the processing SSP set only to low-level API. For high-level API, drain should be solely handled by Operator DAG. Additionally, ask the `TaskCoordinator` to commit the task once all streams for a task have drained.

3. To fix the issue of pipeline getting stuck if a drain control message was present prior to container start, drain and watermark control messages are written to the in-memory buffer on `SystemConsumers` start if it is already is in drain mode.

# Tests:
- End to end tests for both high-level and low-level API
- Unit tests for Drain in RunLoop
- Integ test for Drain for High-level and Low-Level API- DrainLowLevelApiIntegrationTest & DrainHighLevelApiIntegrationTest

# API changes:
None

### CODE REVIEW

File: samza-test/src/test/java/org/apache/samza/test/drain/DrainHighLevelApiIntegrationTest.java [hunk 1]
```diff
@@ -74,20 +74,30 @@ public void describe(StreamApplicationDescriptor appDescriptor) {
...
+   * In order to simulate the real-world behaviour of drain, the test adds messages to the in-memory topic buffer periodically
+   * in a delayed fashion instead of all at once. The test then writes the drain notification message to the in-memory
+   * metadata store to drain and stop the pipeline. This write is done shortly after the pipeline starts and before all
+   * the messages are written to the topic's buffer. As a result, the total count of the processed messages will be less
+   * than the expected count of messages.
+   * */
   @Ignore
   @Test
-  public void testPipeline() {
+  public void testDrain() {
```
Comment (xinyuiscool) [inline samza-test/src/test/java/org/apache/samza/test/drain/DrainHighLevelApiIntegrationTest.java:95]: Seems most of the integration tests are ignored due to relying on wall-clock time. Is it possible to enable a couple and we don't need to do this timed wait? Not sure this TestRunner support waitForFinish(). but samza runners have this api so in theory we should be able to use it.

File: samza-core/src/main/java/org/apache/samza/container/RunLoop.java [hunk 2]
```diff
@@ -875,49 +876,40 @@ private boolean shouldDrain() {
...
-              // from task's processing set irrespective of keyBucket.
-              SystemStreamPartition sspOfEnvelope = envelope.getSystemStreamPartition();
-              Optional<SystemStreamPartition> ssp = processingSspSetToDrain.stream()
-                  .filter(sspInSet -> sspInSet.getSystemStream().equals(sspOfEnvelope.getSystemStream())
-                      && sspInSet.getPartition().equals(sspOfEnvelope.getPartition()))
-                  .findFirst();
-              ssp.ifPresent(processingSspSetToDrain::remove);
-            }
 
             if (!hasIntermediateStreams) {
```
[the remark this answers is not in the document]
Comment (ajothomas) [reply samza-core/src/main/java/org/apache/samza/container/RunLoop.java]: Changed this.

File: samza-core/src/main/java/org/apache/samza/container/RunLoop.java [hunk 3]
```diff
@@ -875,49 +876,40 @@ private boolean shouldDrain() {
...
-              // intermediate streams
+              // The flow below only applies to samza low-level API
+
+              // For high-level API, we do not remove the message from pending queue.
+              // It will be picked by the process flow instead of drain flow, as we want the drain control message
+              // to be processed by the High-level API Operator DAG.
+
+              processingSspSetToDrain.remove(envelope.getSystemStreamPartition());
               pendingEnvelopeQueue.remove();
+              return processingSspSetToDrain.isEmpty();
```
[the remark this answers is not in the document]
Comment (ajothomas) [reply samza-core/src/main/java/org/apache/samza/container/RunLoop.java]: Changed it.

File: samza-test/src/test/java/org/apache/samza/test/drain/DrainHighLevelApiIntegrationTest.java (hunk 1 above)
Comment (ajothomas) [reply samza-test/src/test/java/org/apache/samza/test/drain/DrainHighLevelApiIntegrationTest.java:120]: Tests have been turned on now after a few tweaks.

**Pull request merged**